File: lib/opencoinage/currency.rb
module OpenCoinage
  ##
  # A digital currency contract.
  class Currency
    RDF_TYPE = Vocabulary[:Currency]

    ##
    # Initializes the currency.
    #
    # @param  [RDF::URI, String, #to_s] uri
    #   a valid URI identifying the currency
    # @param  [Hash{Symbol => Object}]  options
    # @option options [String] :name   (nil)
    #   the name of the currency
    # @option options [Issuer] :issuer (nil)
    #   the issuer of the currency
    def initialize(uri = nil, options = {})
      @uri    = RDF::URI(uri || UUID.generate(:urn))
      @name   = options[:name]   if options.has_key?(:name)
      @issuer = options[:issuer] if options.has_key?(:issuer)
    end

    ##
    # The URI identifying this currency.
    #
    # @return [RDF::URI]
    attr_reader :uri
    alias_method :to_uri, :uri

    ##
    # The name of this currency.
    #
    # @return [String]
    # @see    http://xmlns.com/foaf/spec/#term_name
    attr_accessor :name

    ##
    # The issuer of this currency.
    #
    # @return [Issuer]
    # @see    http://opencoinage.org/rdf/issuer
    attr_accessor :issuer

    ##
    # Returns the RDF representation of this currency.
    #
    # @param  [Hash{Symbol => Object}] options
    # @option options [RDF::Resource]  :context (nil)
    # @return [RDF::Graph]
    # @see    http://en.wikipedia.org/wiki/Resource_Description_Framework
    # @see    http://rdf.rubyforge.org/RDF/Graph.html
    def to_rdf(options = {})
      RDF::Graph.new(options[:context]) do |graph|
        graph << [uri, RDF.type, self.class.const_get(:RDF_TYPE)]
        graph << [uri, RDF::FOAF.name, name] if name
        graph << [uri, Vocabulary[:issuer], issuer.to_uri] if issuer
      end
    end
  end # Currency
end # OpenCoinage

File: src/java/org/opencoinage/Currency.java
package org.opencoinage;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.Map;

/**
 * A digital currency contract.
 */
public class Currency {
  public static final String RDF_TYPE = Vocabulary.BASE + "Currency";

  /**
   * The URI identifying this currency.
   *
   * @see java.net.URI
   */
  public final URI uri;

  /**
   * The issuer of this currency.
   */
  public Issuer issuer;

  /**
   * @param uri a valid URI identifying the currency
   */
  public Currency(final URI uri) {
    this.uri = uri;
  }

  /**
   * @param uri a valid URI identifying the currency
   * @param options key-value mappings for fields
   */
  public Currency(final URI uri, final Map<String, ?> options) {
    this.uri = uri;
    if (options != null) {
      if (options.containsKey("issuer")) {
        this.issuer = (Issuer)options.get("issuer"); 
      }
    }
  }

  /**
   * @param uri a valid URI identifying the currency
   * @throws URISyntaxException if <code>uri</code> is not a valid URI
   */
  public Currency(final String uri) throws URISyntaxException {
    this(new URI(uri));
  }

  /**
   * @param uri a valid URI identifying the currency
   * @param options key-value mappings for fields
   * @throws URISyntaxException if <code>uri</code> is not a valid URI
   */
  public Currency(final String uri, final Map<String, ?> options) throws URISyntaxException {
    this(new URI(uri), options);
  }
}
